### PR TITLE
Rianfowler/sc 58965/add typechecking to AppConfig

### PR DESCRIPTION
**Pull request by rianfowler**
#### What this PR does / why we need it:
- refactor: adds typechecking to config page 

#### Which issue(s) this PR fixes:
https://app.shortcut.com/replicated/story/58965/adding-typechecking-to-top-level-routes-routes-in-the-navbar

#### Does this PR introduce a user-facing change?
NONE - refactor that adds type checking 

#### Does this PR require documentation?
NONE

### CODE REVIEW

File: web/src/types/index.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
+export type App = {
```
Comment (rianfowler) [inline web/src/types/index.ts:1]: I moved the app type into here. There are a few references that needed to be updated.

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 2]
```diff
@@ -590,31 +648,23 @@ class AppConfig extends Component {
           </SideNavWrapper>
           <div className="ConfigArea--wrapper">
             <UseIsHelmManaged>
-              {({ data = {} }) => {
-                const { isHelmManaged } = data;
-
-                const {
-                  mutate: saveConfig,
-                  isLoading: isSaving,
-                  isError: saveError,
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:599]: had to remove some unused variables.

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 3]
```diff
@@ -590,31 +648,23 @@ class AppConfig extends Component {
...
-                } = useSaveConfig({
-                  appSlug: this.getSlug(),
-                });
+              {({ data = {} }: { data: { isHelmManaged?: boolean } }) => {
+                const { isHelmManaged: isHelmManagedFromHook } = data;
 
-                const {
-                  download,
-                  clearError: clearDownloadError,
-                  error: downloadError,
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:607]: this never returns an error. i think it did previously but i updated the fetch and didn't have typechecking in place to complain. added a todo in the `useDownloadValues` hook to fix this

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 4]
```diff
@@ -670,26 +720,26 @@ class AppConfig extends Component {
...
                           downloadClicked={download}
-                          downloadError={downloadError}
-                          isDownloading={isDownloading}
                           hideHelmDeployModal={() => {
                             this.setState({ showHelmDeployModal: false });
                             clearDownloadError();
                           }}
                           registryUsername={
-                            this.props?.app?.credentials?.username
+                            this.props?.app?.credentials?.username || ""
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:728]: Could probably remove the optional chaining here but I think we should continue typing App out some more and then figure out a way to initialize it with sane defaults.

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 5]
```diff
@@ -781,4 +831,8 @@ class AppConfig extends Component {
   }
 }
 
-export default withRouter(AppConfig);
+// TODO: fix this type
+// eslint-disable-next-line
+const AppConfigWithRouter: any = withRouter(AppConfig);
+
+export default AppConfigWithRouter;
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:838]: This is a mess.  I need to dig into this to figure out how to type components decorated with react router

File: web/webpack.config.js [hunk 6]
```diff
@@ -49,6 +49,7 @@ module.exports = function (env) {
         "@src": path.resolve(__dirname, "src"),
         "@features": path.resolve(__dirname, "src/features"),
         "@stores": path.resolve(__dirname, "src/stores"),
+        "@types": path.resolve(__dirname, "src/types/index"),
```
Comment (rianfowler) [inline web/webpack.config.js:52]: will need to update these aliases if we decide to break the types file up. for now i think it'll be fine to shove everything into index.ts

File: web/src/features/AppConfig/components/AppConfig.tsx (hunk 5 above)
Comment (miaawong) [reply web/src/features/AppConfig/components/AppConfig.tsx:838]: don't know if this is helpful but from researching, it seems like you can use RouteComponentProps https://www.reactandtypescript.dev/examples/react-router#withrouter

File: web/src/types/index.ts (hunk 1 above)
Comment (miaawong) [reply web/src/types/index.ts:1]: I like it! Makes it more reusable.

**Review by miaawong — APPROVED** [review]

**Pull request merged**